### PR TITLE
fix(validation): replace ancestor-walk with parent-dir validation in require_exists=false branch

### LINKED ISSUE #1134: bug(edit_overwrite): internal_error at <=2ms duration indicates pre-I/O path guard regression since Jun 15
## Summary

`edit_overwrite` fails with `internal_error` at 35–57% since v0.17.0 (released 2026-06-15 03:35 UTC). All errors complete in ≤6 ms. Code analysis confirms failure occurs inside `validate_path_in_dir` (`validation.rs:192–232`) before any file I/O reaches `edit_overwrite_content`.

## Data

| Date | Errors | Calls | Rate | Release in production |
|------|--------|-------|------|-----------------------|
| Jun 14 | 0 | 900 | 0% | v0.16.4 (baseline) |
| Jun 15 | 768 | 1,344 | 57% | v0.17.0 (03:35 UTC) |
| Jun 16 | 972 | 2,016 | 48% | v0.17.x patches |
| Jun 17 | 120 | 1,140 | 11% | v0.18.x / v0.19.0 / v0.20.0 |
| Jun 18 | 228 | 660 | 35% | unresolved |

All 2,088 errors have `duration_ms` ≤ 6 ms, ruling out file I/O, disk, or encoding as the cause.

## Root Cause

`validate_path_in_dir` (`crates/aptu-coder/src/validation.rs:148–253`) has a defect in its `require_exists=false` branch (lines 192–232), which is the path taken when `edit_overwrite` creates a new file.

Defect sequence:

1. The loop at lines 200–215 walks ancestor directories upward from the target path, collecting suffix components, until it finds an ancestor where `canonical_working_dir.join(&ancestor).exists()` is true.
2. When no real ancestor is found, `ancestor` is left as `PathBuf::new()`. The condition becomes `canonical_working_dir.join(PathBuf::new()).exists()`, which is equivalent to `canonical_working_dir.exists()` -- always true. The loop exits without having validated any component of the target path.
3. Line 229: `std::fs::canonicalize(&canonical_base).unwrap_or_else(|e| { warn!(...); canonical_working_dir.clone() })` -- if `canonicalize` fails, the result silently falls back to `canonical_working_dir`. This leaves `canonical_base` pointing at `working_dir` itself, not the intended parent.
4. Line 230 joins `suffix` (collected from unvalidated components) onto this wrong base. The resulting path is not confirmed to be within `working_dir`.
5. `write_file_atomic` in `crates/aptu-coder-core/src/edit.rs` fails on this malformed path; the I/O error surfaces as `internal_error`.

## Partial Fix Already Shipped

Commit `c5c9c29` (released in v0.18.x, Jun 16) added the `warn!()` log on line 229 when `canonicalize` fails. It did not fix the root issue: the `unwrap_or_else` fallback remains, and the suffix components are still joined onto the unvalidated base.

## Affected Code

- `crates/aptu-coder/src/validation.rs:192–232` -- `require_exists=false` branch: ancestor walk, `canonicalize` fallback, suffix join
- `crates/aptu-coder/src/lib.rs:2682` -- `edit_overwrite` handler calls `validate_path_in_dir` (full handler: lines 2651–2750)
- `crates/aptu-coder/src/lib.rs:3006` -- `edit_replace` handler also calls `validate_path_in_dir` (full handler: lines 2960–3050)
- `crates/aptu-coder-core/src/edit.rs` -- `edit_overwrite_content` and `edit_replace_content` receive the malformed path and fail

## Fix

Restructure the `require_exists=false` branch in `validate_path_in_dir` (lines 192–232):

**Preferred approach:** validate the parent directory of the target path using the `require_exists=true` path (which canonicalizes correctly), then append the final filename component. This avoids the ancestor-walk entirely and is safe for new-file creation.

**Alternative:** on each iteration of the ancestor walk, propagate a typed error (`ValidationError::CanonicalizeFailure`) when `canonicalize` fails instead of falling back to `canonical_working_dir`. Remove the `unwrap_or_else` on line 229.

Either approach must ensure that `canonical_base` is confirmed to be within `canonical_working_dir` before `suffix` is joined onto it.

## Test Gaps

- `crates/aptu-coder/tests/edit_working_dir.rs:47–85` (`test_edit_overwrite_new_file_no_working_dir`) covers single-level new file creation only -- does not test multi-level non-existent parent paths (e.g. `working_dir/a/b/c.txt` where `a/b/` does not exist)
- No test for the `ancestor=PathBuf::new()` case (loop exits without finding a real ancestor)
- No test for `working_dir` valid but immediate parent directory does not exist
- No negative test for suffix components containing symlinks that escape `working_dir` after the `unwrap_or_else` fallback

## Acceptance Criteria

- `edit_overwrite` `internal_error` rate returns to 0% (Jun 14 baseline)
- `validate_path_in_dir` `require_exists=false` branch returns a typed error when `canonicalize` fails; no `unwrap_or`/`unwrap_or_else` fallback to `canonical_working_dir`
- New integration tests cover: multi-level non-existent parent creation, `PathBuf::new()` ancestor edge case, and symlink escape attempt in suffix

### PR DESCRIPTION
**Pull request by clouatre**
## Summary

Replaces the defective ancestor-walk loop in the `require_exists=false` branch of both `validate_path` and `validate_path_in_dir` with parent-directory validation, extracted into a shared `validate_parent_in_root` helper.

The old code left `ancestor` as `PathBuf::new()` when no real ancestor was found, causing `canonical_working_dir.join(PathBuf::new()).exists()` (always true) to exit the loop prematurely. The `unwrap_or_else` fallback on `canonicalize` then silently left `canonical_base` pointing at `working_dir` itself, and the unvalidated suffix was joined onto that wrong base, producing a malformed path that `write_file_atomic` rejected as `internal_error`.

New approach (shared by both functions via `validate_parent_in_root`):
1. Extract `file_name` from path; reject with `INVALID_PARAMS` if `None` (path ends with `..` or `/`)
2. Extract parent; if empty or `.` use `root` directly
3. Canonicalize parent; verify `starts_with(root)` and `is_dir()`
4. Join parent + `file_name`; verify final path still `starts_with(root)`

Deeply nested non-existent paths (e.g. `a/b/c/new.txt` where only `a/` exists) are now rejected at validation time with a clear `INVALID_PARAMS` message rather than reaching `write_file_atomic` and failing with an opaque `internal_error`. This is a correctness improvement: `write_file_atomic` has always required the immediate parent to exist; the old code never delivered on an implicit promise to create intermediate directories.

Also aligns all "path is outside the allowed root" error messages to "path is outside the working directory" for consistency.

## Changes

- `crates/aptu-coder/src/validation.rs` -- extract `validate_parent_in_root` helper; replace ancestor-walk loop in both `validate_path` and `validate_path_in_dir`; align error message wording; add comment explaining absolute-path rejection via `starts_with`
- `crates/aptu-coder/tests/edit_working_dir.rs` -- 5 new integration tests: missing parent dir, traversal path, deeply nested path, symlink parent, parent-is-file
- `crates/aptu-coder/src/lib.rs` -- update unit tests to match new validation semantics; simplify `test_validate_path_creates_parent_for_nonexistent_file` setup

## Test plan

- [x] 140 tests pass (`cargo test`)
- [x] Linter clean (`cargo clippy -- -D warnings`)
- [x] Deny clean (`cargo deny check advisories licenses`)
- [x] All inline review comments resolved

Closes #1134

### CODE REVIEW

File: crates/aptu-coder/src/validation.rs [hunk 1]
```diff
@@ -193,47 +196,88 @@ pub(crate) fn validate_path_in_dir(
...
-                // use working_dir as anchor; starts_with below enforces the boundary.
-                ancestor = std::path::PathBuf::new();
-                break;
-            }
+        // (a) Reject paths where file_name is None (bare '..', '.', or trailing slash).
+        let file_name = p.file_name().ok_or_else(|| {
+            ErrorData::new(
+                rmcp::model::ErrorCode::INVALID_PARAMS,
+                "path must include a filename component".to_string(),
+                Some(error_meta(
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs]: Done. Extracted `validate_parent_in_root(path, root)` as a private helper and updated both `validate_path` and `validate_path_in_dir` to call it. The ancestor-walk loop is removed from `validate_path` as well.

File: crates/aptu-coder/src/validation.rs [hunk 2]
```diff
@@ -193,47 +196,88 @@ pub(crate) fn validate_path_in_dir(
...
+        };
+
+        // (c) Canonicalize parent (require_exists=true logic, inline).
+        let canonical_parent = std::fs::canonicalize(&parent_path).map_err(|e| {
+            io_error_to_path_error(
+                &e,
+                parent.to_str().unwrap_or("(invalid utf-8)"),
+                "provide a valid parent directory within the working directory",
+            )
+        })?;
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs]: Fixed. All occurrences of "path is outside the allowed root" in `validate_path` are now "path is outside the working directory", consistent with the wording used throughout `validate_path_in_dir`.

File: crates/aptu-coder/src/validation.rs [hunk 3]
```diff
@@ -3,10 +3,95 @@
...
+        root.to_path_buf()
+    } else {
+        root.join(parent)
+    };
+
+    // Canonicalize parent.
+    let canonical_parent = std::fs::canonicalize(&parent_path).map_err(|e| {
+        io_error_to_path_error(
+            &e,
+            parent.to_str().unwrap_or("(invalid utf-8)"),
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:46]: Good catch. Absolute paths are rejected by the `starts_with(root)` check on line 53: if an absolute path is passed in, `root.join(parent)` returns the absolute path as-is (Rust's `Path::join` discards the base when the argument is absolute), so `canonicalize` resolves it, and then the boundary check rejects it since it won't start with `root`. Added a comment in the code to document this behaviour explicitly.

File: crates/aptu-coder/src/validation.rs [hunk 4]
```diff
@@ -3,10 +3,95 @@
...
+            "provide a valid parent directory within the working directory",
+        )
+    })?;
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:49]: Acknowledged. `OsStr` is not directly `Display`able, so `to_str().unwrap_or("(invalid utf-8)")` is idiomatic for user-facing error messages where a UTF-8 path is expected. For structured debug logging we already capture the raw `io::Error` kind and message in the `ioErrorKind`/`ioErrorSource` fields of the error metadata, which gives enough context without requiring a separate OsStr representation.

File: crates/aptu-coder/src/lib.rs [hunk 5]
```diff
@@ -5024,15 +5024,23 @@ mod tests {
 
     #[test]
     fn test_validate_path_creates_parent_for_nonexistent_file() {
-        // Edge case: non-existent file with non-existent parent should still be accepted
-        // if the ancestor chain leads back to CWD.
-        let result = validate_path("nonexistent_dir/nonexistent_file.txt", false);
+        // Edge case: non-existent file with existing parent should be accepted.
+        let cwd = std::env::current_dir().expect("should get cwd");
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/lib.rs:5028]: Done. Simplified to `parent.path().to_path_buf()` and `parent_path.join("new_file.txt")`. The `file_name()` extraction and string formatting are gone.

File: crates/aptu-coder/src/lib.rs [hunk 6]
```diff
@@ -5102,18 +5110,11 @@ mod tests {
         // Act: try to traverse outside working_dir with ../../../etc/passwd
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/lib.rs:5106]: Acknowledged. The test name and assertion message have been updated to reflect the new behaviour (parent must exist; deeply nested non-existent paths are correctly rejected).

File: crates/aptu-coder/src/validation.rs [hunk 7]
```diff
@@ -29,7 +118,7 @@ pub(crate) fn validate_path(
     let allowed_root = std::fs::canonicalize(&cwd).map_err(|_| {
         ErrorData::new(
             rmcp::model::ErrorCode::INVALID_PARAMS,
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:120]: Already addressed. The `unwrap_or_else` fallback on `canonicalize` was removed as part of the ancestor-walk refactor; `validate_path` now propagates the error via `?` through `io_error_to_path_error`. No silent fallback remains.

File: crates/aptu-coder/src/lib.rs [hunk 8]
```diff
@@ -5336,25 +5333,13 @@ mod tests {
     #[test]
     fn test_validate_path_in_dir_nonexistent_deep_path() {
         // Deeply nested non-existent path: a/b/c/d/new.txt -- none of the
-        // intermediate directories exist.  The loop must walk up all four
-        // segments and anchor at working_dir, then rejoin the full suffix.
+        // intermediate directories exist.  With parent-directory validation,
+        // this is rejected because the parent a/b/c/d does not exist.
         let temp_dir = tempfile::TempDir::new().expect("should create temp dir");
         let result = validate_path_in_dir("a/b/c/d/new.txt", false, temp_dir.path());
         assert!(
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/lib.rs:5340]: Not a breaking change. `edit_overwrite` has never created intermediate directories -- `write_file_atomic` requires the parent to exist and returns an OS error if it does not. With the old ancestor-walk the failure mode was an opaque `internal_error` at write time (the loop produced a malformed path that the filesystem rejected). With the new validation the same call returns `INVALID_PARAMS` with a clear message at validation time, before any write is attempted. The observable contract (create a file whose parent exists) is unchanged; only the error site and message improved. No documentation of a breaking change is needed.
